providers/byteplus/byteplusProvider.go

```diff
@@ -0,0 +1,347 @@
+		// they seem to manage them on deletes and things, so we'll just pretend they don't exist
+		if rtype == "TXT" && strings.HasPrefix(rcontent, "ALIAS for ") {
+			continue
+		}
+
+		rc := &models.RecordConfig{
+			Original: record,
+		}
+		if record.TTL != nil {
+			rc.TTL = uint32(*record.TTL)
+		}
+		rc.SetLabel(rname, domainName)
+
+		switch rtype {
+		case "ALIAS", "URL":
+			rc.Type = rtype
+			rc.SetTarget(rcontent)
+		case "MX":
+			sp := strings.Split(*record.Value, " ")       // received combined value from byteplus "5 domain.com"
+			rcontent = strings.Join(sp[1:], " ") + "."    // re-add trailing dot "5 domain.com."
+			rprio, err := strconv.ParseInt(sp[0], 10, 64) // split get priority value
+
+			var prio uint16
+			if err != nil {
+				return nil, err
+			}
+			prio = uint16(rprio)
+
+			rc.SetTargetMX(prio, rcontent)
+		default:
+			err = rc.PopulateFromString(rtype, rcontent, domainName)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("unparsable record received from byteplus: %w", err)
+		}
+
+		existingRecords = append(existingRecords, rc)
+	}
+
+	return existingRecords, nil
+}
+
+// GetZoneRecordsCorrections returns a list of corrections that will turn existing records into dc.Records.
+func (c *byteplusProvider) GetZoneRecordsCorrections(dc *models.DomainConfig, actual models.Records) ([]*models.Correction, error) {
+	corrections, err := c.getDiff2DomainCorrections(dc, actual)
+	if err != nil {
+		return nil, err
+	}
+
+	return corrections, nil
+}
+
+func (c *byteplusProvider) getDiff2DomainCorrections(dc *models.DomainConfig, actual models.Records) ([]*models.Correction, error) {
+	domain, err := c.findDomainByName(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	var corrections []*models.Correction
+	instructions, err := diff2.ByRecord(actual, dc, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, inst := range instructions {
+		switch inst.Type {
+		case diff2.REPORT:
+			corrections = append(corrections, &models.Correction{Msg: inst.MsgsJoined})
+		case diff2.CHANGE:
+			corrections = append(corrections, &models.Correction{
+				Msg: inst.Msgs[0],
+				F:   c.updateRecordFunc(inst.Old[0].Original.(*byteplus.QueryRecordResponse), inst.New[0], domain.ZID),
+			})
+		case diff2.CREATE:
+			corrections = append(corrections, &models.Correction{
+				Msg: inst.Msgs[0],
+				F:   c.createRecordFunc(inst.New[0], domain.ZID),
+			})
+		case diff2.DELETE:
+			rec := inst.Old[0].Original.(*byteplus.QueryRecordResponse)
+			corrections = append(corrections, &models.Correction{
+				Msg: inst.Msgs[0],
+				F:   c.deleteRecordFunc(*rec.RecordID),
+			})
+		default:
+			panic(fmt.Sprintf("unhandled inst.Type %s", inst.Type))
+		}
+	}
+	return corrections, nil
+}
+
+// Returns a function that can be invoked to create a record in a zone.
+func (c *byteplusProvider) createRecordFunc(rc *models.RecordConfig, domainID *int64) func() error {
+	return func() error {
+		target := rc.GetTargetCombined()
```

Comment:
There's a lot of code duplication in createRecordFunc and updateRecordFunc.  Suggestion: Move the common code to a separate function.